method/message/message.json

```diff
@@ -0,0 +1,67 @@
+{
+    "$comment": "By executing a publish action, an attendee communicates its intention to publish a specific message on a channel.",
+    "$schema": "http://json-schema.org/draft/2019-09/schema#",
+    "$id": "http://pop2020.dedis.ch/message.json",
+    "type": "object",
+    "properties" : {
+        "jsonrpc": {
+            "const": "2.0"
+        },
+        "method": {
+            "description": "Operation to be performed",
+            "type": "string",
+            "enum": ["publish", "broadcast"]
+        },
+        "params": {
+            "description": "Content of the request",
+            "type": "object",
+            "properties": {
+                "channel": {
+                    "description": "Channel id",
+                    "type": "string"
+                },
+                "message": {
+                    "description": "Content of the message",
+                    "type": "object",
+                    "properties": {
+                        "data": {
+                            "description": "base64 representation of the data",
+                            "type": "object",
+                            "$comment": "must be in base64"
+                        },
+                        "sender": {
+                            "description": "Public key of sender",
+                            "type": "string"
+                        },
+                        "signature": {
+                            "description": "Signature by sender over data",
+                            "type": "string"
+                        },
+                        "message_id": {
+                            "$comment": "BFord: Not sure that this actually needs to be transmitted, since it can simply be computed from the data and signature?",
+                            "description": "hash(data||signature)",
+                            "type": "string"
+                        },
+                        "witness_signatures": {
+                            "description": "Signature by witnesses (sender||data)",
+                            "type": "array",
+                            "items": {
+                                "$comment": "slightly unsure about the type here",
+                                "$comment": "initially empty array, will be filled on further message. Not 100% sure whether it should be in the required list (but allowed to be empty) or not",

```

Comment:
can't have two fields with the same name :) 

type should probably be string, you're right. (possibly base64 encoded signatures?)

Reply:
In this particular case, it's easy to solve, but generally, is the two fields rule an issue even with $comment? 

Because my understanding of $comment in json-schema is that those fields are not considered as belonging to the actual json.
And multi-line comments are also impossible, so that was my alternative to make them more human readable.

Reply:
JSON is bad for humans :-P

This being said, I understand your concern. It's not too much of an issue with `$comment` fields, but be careful about this with other fields as the final value will only be that of the last defined field.